exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/FunctionImplementationRegistry.java

```diff
@@ -301,29 +323,120 @@ private ScanResult scan(ClassLoader classLoader, Path path, URL[] urls) throws I
         return RunTimeScan.dynamicPackageScan(drillConfig, Sets.newHashSet(urls));
       }
     }
-    throw new FunctionValidationException(String.format("Marker file %s is missing in %s.",
+    throw new JarValidationException(String.format("Marker file %s is missing in %s",
         CommonConstants.DRILL_JAR_MARKER_FILE_RESOURCE_PATHNAME, path.getName()));
   }
 
-  private static String getUdfDir() {
-    return Preconditions.checkNotNull(System.getenv("DRILL_UDF_DIR"), "DRILL_UDF_DIR variable is not set");
+  /**
+   * Return list of jars that are missing in local function registry
+   * but present in remote function registry.
+   *
+   * @param remoteFunctionRegistry remote function registry
+   * @param localFunctionRegistry local function registry
+   * @return list of missing jars
+   */
+  private List<String> getMissingJars(RemoteFunctionRegistry remoteFunctionRegistry,
+                                      LocalFunctionRegistry localFunctionRegistry) {
+    List<Jar> remoteJars = remoteFunctionRegistry.getRegistry().getJarList();
+    List<String> localJars = localFunctionRegistry.getAllJarNames();
+    List<String> missingJars = Lists.newArrayList();
+    for (Jar jar : remoteJars) {
+      if (!localJars.contains(jar.getName())) {
+        missingJars.add(jar.getName());
+      }
+    }
+    return missingJars;
+  }
+
+  /**
+   * Creates local udf directory, if it doesn't exist.
+   * Checks if local is a directory and if current application has write rights on it.
+   * Attempts to clean up local idf directory in case jars were left after previous drillbit run.
+   *
+   * @return path to local udf directory
+   */
+  private Path getLocalUdfDir() {
+    String confDir = getConfDir();
```

Comment:
Unfortunately, I didn't realize that $DRILL_HOME and $DRILL_CONF_DIR in DoY are not writable.
Yes, you are right we do clean up local udf directory from previously loaded jars each time drillbit startups so it definitely a good idea to keep local udf directory in tmp folder,
So I have added $DRILL_TMP_DIR in drill-config.sh with default to /tmp if not set.
In code I concatenate $DRILL_TMP_DIR + drill.exec.udf.directory.base which defaults to ${drill.exec.zk.root}"/udf".